photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/AbstractOptimizer.scala

```diff
@@ -44,7 +44,17 @@ abstract class AbstractOptimizer[Datum <: DataPoint, -Function <: DiffFunction[D
     None
   }
 
-  def isDone: Boolean = convergenceReason.nonEmpty
+  def isDone: Boolean = {
+    val isDone = convergenceReason.nonEmpty
+    if (isDone) {
+      // Set the convergenceReason when the optimization is done
+      statesTracker match {
+        case Some(y) => y.convergenceReason = convergenceReason
+        case None =>
+      }
+    }
+    isDone
+  }
```

Comment:
Good suggestion.

I put this piece of code outside of ```setCurrentState``` because ```isDone``` depends on the state of ```previousState```, and coupling ```previousState``` with ```setCurrentState``` might be tricky because the ```Optimizer``` class can change the state of ```previousState``` without informing ```currentState```.

Let me see what I can do to simply the logic a little bit without introducing additional side effect.